rust/connlib/tunnel/src/client.rs

```diff
@@ -171,18 +171,35 @@ where
         Ok(())
     }
 
-    /// Sets the interface configuration and starts background tasks.
-    #[tracing::instrument(level = "trace", skip(self))]
-    pub fn set_interface(&mut self, config: &InterfaceConfig) -> connlib_shared::Result<()> {
-        self.role_state.interface_config = Some(config.clone());
-        let effective_dns_servers = effective_dns_servers(
-            config.upstream_dns.clone(),
-            self.callbacks
-                .get_system_default_resolvers()
-                .ok()
-                .flatten()
-                .unwrap_or_default(),
-        );
+    pub fn set_dns(&mut self, new_dns: Vec<IpAddr>) {
+        if self
+            .role_state
+            .system_resolvers
+            .as_ref()
+            .is_some_and(|old_dns| {
+                HashSet::<&IpAddr>::from_iter(old_dns.iter())
+                    == HashSet::<&IpAddr>::from_iter(new_dns.iter())
+            })
+        {
+            return;
+        }
+
+        self.role_state.system_resolvers = Some(new_dns);
+        if let Err(err) = self.update_interface() {
+            tracing::error!("Failed to update interface: {err:?}");
+        }
+    }
+
+    fn update_interface(&mut self) -> connlib_shared::Result<()> {
+        let Some(config) = self.role_state.interface_config.as_ref().cloned() else {
+            return Ok(());
+        };
+        let Some(system_resolvers) = self.role_state.system_resolvers.as_ref().cloned() else {
+            return Ok(());
+        };
+
+        let effective_dns_servers =
+            effective_dns_servers(config.upstream_dns.clone(), system_resolvers);
```

Comment:
Isn't this duplicated state? If our effective DNS servers contain the system resolves, why can't we just check our current DNS mapping instead of tracking the system resolvers separately?

We are already tracking so much state, we should avoid redundancy as much as we can.

Reply:
because, if we have some upstream dns the effective dns doesn't contain the system's resolver

Reply:
Can we unit test this somehow? Seems fragile to me :)

Reply:
We need to stop reaching into fields of `ClientState` from the `Tunnel` and make proper APIs that manipulate `ClientState`.

Reply:
> Can we unit test this somehow? Seems fragile to me :)

I can add some unit-tests for `effective_dns_servers` but is that what you want to unit-test?

> We need to stop reaching into fields of ClientState from the Tunnel and make proper APIs that manipulate ClientState.

We can add a `ClientState::get_dns` that returns this `effective_dns_servers` since both part of this function is stored there, what do you thinka bout that?

Reply:
What I'd like to test for example is that setting the same DNS servers is a no-op.

For example, we could make a function that given a set of DNS servers outputs an `Option` of `dns_mapping` and only update `Io` if that is `Some`. Maybe there is more stuff we can do, just need to set down and de-couple the side-effects from the state management and translate it into values that we pass around and then can write tests against.